Source/SharpNeedle/HedgehogEngine/Mirage/SplinePath.cs

```diff
@@ -0,0 +1,283 @@
+        writer.WriteArrayOffset(KnotFlags.ToArray(), 4);
+        writer.WriteArrayOffset(KnotDistances.ToArray(), 4);
+        writer.WriteArrayOffset(Knots.ToArray(), 4);
+        writer.WriteArrayOffset(KnotTangentIns.ToArray(), 4);
+        writer.WriteArrayOffset(KnotTangentOuts.ToArray(), 4);
+
+        writer.Write(DoubleKnots.Count * 2);
+        if (DoubleKnots.Count != 0)
+            writer.WriteObjectCollectionOffset(DoubleKnots);
+        else
+            writer.Write<int>(0);
+
+        writer.Write(Aabb);
+
+        writer.Write(UserDatas.Count);
+        writer.WriteObjectCollectionOffset(UserDatas);
+
+        writer.Write(Field48);
+
+        writer.WriteObjectOffset(Unknown);
+    }
+
+    public override string ToString()
+    {
+        return Name;
+    }
+}
+
+public class DoubleKnot : IBinarySerializable
+{
+    public Vector3 Left { get; set; }
+    public Vector3 Right { get; set; }
+
+    public void Read(BinaryObjectReader reader)
+    {
+        Left = reader.Read<Vector3>();
+        Right = reader.Read<Vector3>();
+    }
+
+    public void Write(BinaryObjectWriter writer)
+    {
+        writer.Write(Left);
+        writer.Write(Right);
+    }
+}
+
+public class UserData : IBinarySerializable
+{
+    public string Name { get; set; }
+    public DataType Type { get; set; }
+    public uint Value { get; set; }
+    public string? StringValue { get; set; }
+
+    public unsafe float FloatValue
+    {
+        get
+        {
+            var v = Value;
+            return *(float*)&v;
+        }
+
+        set => Value = *(uint*)&value;
+    }
+
+    public void Read(BinaryObjectReader reader)
+    {
+        Name = reader.ReadStringOffset();
+
+        Type = reader.Read<DataType>();
+        switch (Type)
+        {
+            case DataType.Float:
+            case DataType.Integer:
+                Value = reader.Read<uint>();
+                break;
+
+            case DataType.String:
+                StringValue = reader.ReadStringOffset();
+                break;
+
+            default:
+                throw new NotImplementedException($"{Type} is not recognized");
+        }
+    }
+
+    public void Write(BinaryObjectWriter writer)
+    {
+        writer.WriteStringOffset(StringBinaryFormat.NullTerminated, Name);
+
+        writer.Write(Type);
+        switch (Type)
+        {
+            case DataType.Integer:
+            case DataType.Float:
+                writer.Write(Value);
+                break;
+
+            case DataType.String:
+                writer.WriteStringOffset(StringBinaryFormat.NullTerminated, StringValue);
+                break;
+
+            default:
+                throw new NotImplementedException($"{Type} is not recognized");
+        }
+    }
+
+    public override string ToString()
+    {
+        return Name;
+    }
+
+    public enum DataType
+    {
+        Integer,
+        Float, // Guessed
+        String
+    }
+}
+
+public class UnknownStruct : IBinarySerializable
```

Comment:
Should move this a subclass of PathObject as `UnknownStruct` can be a very common name.

Reply:
Resolved in 6a814e4.